### DISCUSSION TITLE
what's the proper way to add a Flarum extension?

### DISCUSSION THREAD

**Discussion by PrimateCoder** [shinsenter/php#65] [Q&A]
Hi @shinsenter!

I've tried a couple of ways of adding Flarum extensions to shinsenter/flarum:latest but none of them seem to work properly.

1. I tried to add "RUN composer require <ext>" in the Dockerfile, but if I do this then Flarum doesn't get installed at all, only the extension does.
2. I added a new start up script named 99-add-flarum-extensions.sh, but although this works, the script gets run on every start up and that is unnecessary. I could make the script self-delete, but that is hacky.

What is the proper/official way to add additional Flarum extensions to your image?

Secondly, Flarum complains that certain directories don't have the correct permissions out of the box. I had to add these to my commands:

chmod -R g+w /var/www/html/storage
chmod g+w /var/www/html/public/assets/

How does it work for you?

Thanks so much for the great work!

**Comment by shinsenter** [comment]
Hi @PrimateCoder,

Thank you for reaching out with your question.

Since I don't know the exact content of the project directory or the Dockerfile you are using, it's difficult for me to reproduce and figure out the issue you mentioned.

Therefore, I will assume that you have a directory `www` containing the source code of Flarum and all the necessary packages for your project are declared in the `composer.json` file and a `Dockerfile` for building your Docker image. You then copy the `www` directory into your Docker container, install additional packages using composer, and adjust the permissions for the directories within Flarum's document root. Given this scenario, you might try using the following `Dockerfile` to build your Docker image according to your requirements:

```Dockerfile
FROM shinsenter/flarum:latest

# copy the `www` directory into the container
ENV APP_PATH=/var/www/html
ADD --link --chown="$APP_USER:$APP_GROUP" ./www/ $APP_PATH/

# install additional packages using composer
RUN cd $APP_PATH && composer install

# adjust the permissions for directories
RUN mkdir -p $APP_PATH/html/storage $APP_PATH/public/assets
RUN chmod -R "$APP_USER:$APP_GROUP" $APP_PATH
```

Best regards,
TAN

**Reply by PrimateCoder** [reply]
Thank you very much for the detailed answer!

In the end, I concluded that it was more optimal to start with your PHP Docker image because then I could install flarum and all the extensions at build time rather than runtime. Effectively, I think that's also your suggestion above.

I believe with your current flarum image, by default, flarum is installed only when the container is first started. I think it would be more predictable to handle the installation process when building the image and so that's why I decided to use your excellent PHP Docker images directly.

Thank you!
Navin.
